### PR TITLE
Feature: Item Exchange

### PR DESCRIPTION
**Pull request by Zintixx**
- This is for exchanging fragments( or scrolls) for items.
- Parsing data required for item exchange (fusion) including meso cost, item(s) cost, and exchange item(s)
- Uses the same checks on recipe from MasteryHandler.
- Small fix on MasteryHandler to subtract the correct amount.

Also, not sure what the standard was when it comes to item parsing whether it should be included all in ItemParser.cs or have it's own Parser like how I currently have it.

### CODE REVIEW

File: GameDataParser/Parsers/ItemExchangeScrollParser.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                    ItemExchangeScrollMetadata metadata = new ItemExchangeScrollMetadata();
+
+                    if (node.Name == "exchange")
+                    {
+                        metadata.ExchangeId = int.Parse(node.Attributes["id"].Value);
+                        metadata.Type = node.Attributes["type"].Value;
+                    }
+                    foreach (XmlNode childNode in node)
+                    {
+                        if (childNode.Name == "receipe")
```
Comment (AlanMorel) [inline GameDataParser/Parsers/ItemExchangeScrollParser.cs:36]: typo?

File: GameDataParser/Parsers/ItemExchangeScrollParser.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                            metadata.RewardRarity = byte.Parse(childNode.Attributes["rank"].Value);
+                            metadata.RewardAmount = short.Parse(childNode.Attributes["count"].Value);
+                        }
+                        else if (childNode.Name == "require")
+                        {
+                            metadata.MesoCost = string.IsNullOrEmpty(childNode.Attributes["meso"].Value) ? 0 : int.Parse(childNode.Attributes["meso"].Value);
+
+                            foreach (XmlNode itemNode in childNode)
+                            {
+                                if (itemNode.Name == "item")
```
Comment (AlanMorel) [inline GameDataParser/Parsers/ItemExchangeScrollParser.cs:54]: Invert and continue

File: MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+            if (exchange.ItemCost.Count != 0)
+            {
+                if (!PlayerHasAllIngredients(session, exchange, quantity))
+                {
+                    session.Send(ItemExchangePacket.Notice((short) ExchangeNotice.InsufficientItems));
+                    return;
+                }
+
+                if (RemoveRequiredItemsFromInventory(session, exchange, item, quantity))
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs:82]: Invert and return

File: MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                {
+                    Item exchangeRewardItem = new(exchange.RewardId)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs:84]: 84-91 matches 96-104 almost exactly, maybe do
```InventoryController.Consume(session, item.Uid, exchange.RecipeAmount * quantity);``` in that else, then the rest can be the code where they are identical

File: MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            // TODO: Check if rarity matches
+            
+            List<Item> playerInventoryItems = new(session.Player.Inventory.Items.Values);
+
+            for (int i = 0; i < exchange.ItemCost.Count; i++)
+            {
+                ItemRequirementMetadata exchangeItem = exchange.ItemCost.ElementAt(i);
+                Item item = playerInventoryItems.FirstOrDefault(x => x.Id == exchangeItem.Id);
+
+                if (item != null)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs:119]: Invert and continue

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: GameDataParser/Parsers/ItemExchangeScrollParser.cs (hunk 1 above)
Comment (Zintixx) [reply GameDataParser/Parsers/ItemExchangeScrollParser.cs:36]: yes but not from me

File: MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                    }
+                    int exchangeAmount = exchangeItem.Amount * quantity;
+                    if (exchangeAmount == item.Amount)
+                    {
+                        InventoryController.Remove(session, item.Uid, out Item _);
+                    }
+                    else
+                    {
+                        InventoryController.Update(session, item.Uid, item.Amount - exchangeAmount);
+                    }
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs:139]: Can't use `Consume()` here too?

Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs:139]: oh yeah oops

**Review by Sparkymod — APPROVED** [review]
Nice!

**Review by angelotadeucci — APPROVED** [review]

**Pull request merged**